spread/openstack.go

```diff
@@ -0,0 +1,758 @@
+var openstackProvisionRetry = 5 * time.Second
+
+func (p *openstackProvider) waitProvision(ctx context.Context, s *openstackServer) error {
+	debugf("Waiting for %s to provision...", s)
+
+	timeout := time.After(openstackProvisionTimeout)
+	retry := time.NewTicker(openstackProvisionRetry)
+	defer retry.Stop()
+
+	for {
+		select {
+		case <-timeout:
+			return fmt.Errorf("timeout waiting for %s to provision", s)
+
+		case <-retry.C:
+			server, err := p.computeClient.GetServer(s.d.Id)
+			if err != nil {
+				return fmt.Errorf("cannot get server info: %v", &openstackError{err})
+			}
+			if server.Status != nova.StatusBuild {
+				if server.Status != nova.StatusActive {
+					return fmt.Errorf("cannot use server: status is not active but %s", server.Status)
+				}
+				return nil
+			}
+		case <-ctx.Done():
+			return fmt.Errorf("cannot wait for %s to provision: interrupted", s)
+		}
+	}
+	panic("unreachable")
+}
+
+var openstackServerBootTimeout = 3 * time.Minute
+var openstackServerBootRetry = 5 * time.Second
+
+func (p *openstackProvider) waitServerBootSSH(ctx context.Context, s *openstackServer) error {
+	config := &ssh.ClientConfig{
+		User:            "root",
+		Auth:            []ssh.AuthMethod{ssh.Password(p.options.Password)},
+		Timeout:         10 * time.Second,
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+	addr := s.address
+	if !strings.Contains(addr, ":") {
+		addr += ":22"
+	}
+
+	// Iterate until the ssh connection to the host can be stablished
+	// In openstack the client cannot access to the serial console of the instance
+	timeout := time.After(openstackServerBootTimeout)
+	retry := time.NewTicker(openstackServerBootRetry)
+	defer retry.Stop()
+
+	for {
+		select {
+		case <-timeout:
+			return fmt.Errorf("cannot ssh to the allocated instance: timeout reached")
+		case <-retry.C:
+			_, err := sshDial("tcp", addr, config)
+			if err == nil {
+				debugf("Connection to server %s established", s.d.Name)
+				return nil
+			}
+		case <-ctx.Done():
+			return fmt.Errorf("cannot wait for %s to boot: interrupted", s)
+		}
+	}
+}
+
+func (p *openstackProvider) getSerialConsoleOutput(s *openstackServer) (string, error) {
+	url := fmt.Sprintf("servers/%s/action", s.d.Id)
+
+	var req struct {
+		OsGetSerialConsole struct{} `json:"os-getConsoleOutput"`
+	}
+	var resp struct {
+		Output string `json:"output"`
+	}
+	requestData := goosehttp.RequestData{ReqValue: req, RespValue: &resp, ExpectedStatus: []int{http.StatusOK}}
+
+	err := p.osClient.SendRequest("POST", "compute", "v2", url, &requestData)
+	if err != nil {
+		return "", err
+	}
+
+	return resp.Output, nil
+}
+
+func (p *openstackProvider) waitServerBootSerial(ctx context.Context, s *openstackServer) error {
+	timeout := time.After(openstackServerBootTimeout)
+	relog := time.NewTicker(60 * time.Second)
+	defer relog.Stop()
+	retry := time.NewTicker(openstackServerBootRetry)
+	defer retry.Stop()
+
+	var marker = openstackReadyMarker
+	for {
```

Comment:
@sergiocazzolato  do we need this loop? does the serial console take some time to be ready?

Reply:
sorry my bad, I meant in `getSerialConsoleOutput`, it was a loop before, I updated it.